sky/adaptors/nebius.py

```diff
@@ -98,3 +118,99 @@ def get_tenant_id():
+                             '`sky check` to verify credentials are '
+                             'correctly set up.')
+    return nebius_credentials.get_frozen_credentials()
+
+
+# lru_cache() is thread-safe and it will return the same session object
+# for different threads.
+# Reference: https://docs.python.org/3/library/functools.html#functools.lru_cache # pylint: disable=line-too-long
+@annotations.lru_cache(scope='global')
+def session():
+    """Create an AWS session."""
+    # Creating the session object is not thread-safe for boto3,
+    # so we add a reentrant lock to synchronize the session creation.
+    # Reference: https://github.com/boto/boto3/issues/1592
+    # However, the session object itself is thread-safe, so we are
+    # able to use lru_cache() to cache the session object.
+    with _session_creation_lock:
+        session_ = boto3.session.Session(profile_name=NEBIUS_PROFILE_NAME)
+    return session_
+
+
+@annotations.lru_cache(scope='global')
+def resource(resource_name: str, **kwargs):
+    """Create a Nebius resource.
+
+    Args:
+        resource_name: Nebius resource name (e.g., 's3').
+        kwargs: Other options.
+    """
+    # Need to use the resource retrieved from the per-thread session
+    # to avoid thread-safety issues (Directly creating the client
+    # with boto3.resource() is not thread-safe).
+    # Reference: https://stackoverflow.com/a/59635814
+
+    session_ = session()
+    nebius_credentials = get_nebius_credentials(session_)
+    endpoint = create_endpoint()
+
+    return session_.resource(
+        resource_name,
+        endpoint_url=endpoint,
+        aws_access_key_id=nebius_credentials.access_key,
+        aws_secret_access_key=nebius_credentials.secret_key,
+        region_name='auto',
+        **kwargs)
+
+
+@annotations.lru_cache(scope='global')
+def client(service_name: str, region):
+    """Create an Nebius client of a certain service.
+
+    Args:
+        service_name: Nebius service name (e.g., 's3').
+        kwargs: Other options.
+    """
+    # Need to use the client retrieved from the per-thread session
+    # to avoid thread-safety issues (Directly creating the client
+    # with boto3.client() is not thread-safe).
+    # Reference: https://stackoverflow.com/a/59635814
+
+    session_ = session()
+    nebius_credentials = get_nebius_credentials(session_)
+    endpoint = create_endpoint(region)
+
+    return session_.client(service_name,
+                           endpoint_url=endpoint,
+                           aws_access_key_id=nebius_credentials.access_key,
+                           aws_secret_access_key=nebius_credentials.secret_key,
+                           region_name=region)
+
+
+@common.load_lazy_modules(_LAZY_MODULES)
+def botocore_exceptions():
+    """AWS botocore exception."""
+    # pylint: disable=import-outside-toplevel
+    from botocore import exceptions
+    return exceptions
+
+
+def create_endpoint(region: Optional[str] = DEFAULT_REGION) -> str:
```

Comment:
```suggestion
def create_endpoint(region: str = DEFAULT_REGION) -> str:
```
should this be str? Or, should we handle the case when the argument is None?